### ISSUE TITLE
Rewrite syntax EBNF

### ISSUE THREAD

**Issue by maxonfjvipon** [objectionary/eo#2399]
Current syntax EBNF is quite poor and "weak". There are not enough structure blocks to resolve almost all our syntax errors: #415, #1468, #1469, #1841, #2291, #2320, #2321. 
Need to rewrite it and make stricter

**Comment by maxonfjvipon** [comment]
@Graur please set `Urgent` milestone to it

**Comment by yegor256** [comment]
@maxonfjvipon I didn't understand the problem described here. What exactly is wrong? You are talking about `Program.g4` file?

**Comment by maxonfjvipon** [comment]
@yegor256 yes, about `Program.g4` where `EBNF` is described (if I'm not mistaken in terms). Current amount and structure of key syntax blocks (like `head`, `application`, `htail`) don't allow us to resolve all of above syntax bugs by small changes and fixes because there's a mess in there. I'm sure that current grammar allows us to write much invalid code, we just haven't come across such examples yet. 
I strongly believe that we should make a kind of breakthrough change - rewrite the grammar, make it fuller, wider and from the very beginning to put a solution of all bugs above there

**Comment by yegor256** [comment]
@maxonfjvipon sounds like a good intent, but I'm not sure it's possible to just "make it better" :) You are more than welcome to try.

**Comment by maxonfjvipon** [comment]
@yegor256 image this code:
```
[name] > cat
  name > @

* cat > cats

stdout > @
  cats.at 0
    "Lisa"
```

With current grammar this code is valid but it does not work.
1. I understand it like: "get cat by 0 index (application of `tuple.at` firstly) -> set name to "Lisa"(application of `cat` secondly)"
2. Compiler understands it like: "application of `tuple.at` with to arguments - 0 and "Lisa""

So there's a conflict of vertical and horizontal applications. I believe this code should be invalid on the level of grammar, means we should not allow to mix horizontal and vertical applications even with braces, so this should be not valid too:
```
(cats.at 0)
  "Lisa"
```

What do you think?

**Comment by yegor256** [comment]
@maxonfjvipon this issue does make sense (the one about vertical-vs-horizontal notation), but it should be in a new ticket

**Comment by 0pdd** [comment]
@maxonfjvipon 3 puzzles [#2482](https://github.com/objectionary/eo/issues/2482), [#2483](https://github.com/objectionary/eo/issues/2483), [#2484](https://github.com/objectionary/eo/issues/2484) are still not solved.

**Comment by 0pdd** [comment]
@maxonfjvipon 2 puzzles [#2483](https://github.com/objectionary/eo/issues/2483), [#2484](https://github.com/objectionary/eo/issues/2484) are still not solved; solved: [#2482](https://github.com/objectionary/eo/issues/2482).

**Comment by 0pdd** [comment]
@maxonfjvipon the puzzle [#2483](https://github.com/objectionary/eo/issues/2483) is still not solved; solved: [#2482](https://github.com/objectionary/eo/issues/2482), [#2484](https://github.com/objectionary/eo/issues/2484).

**Comment by 0pdd** [comment]
@maxonfjvipon all 3 puzzles are solved here: [#2482](https://github.com/objectionary/eo/issues/2482), [#2483](https://github.com/objectionary/eo/issues/2483), [#2484](https://github.com/objectionary/eo/issues/2484).
